src/coreComponents/constitutive/fluid/MultiPhaseMultiComponentFluid.hpp

```diff
@@ -267,31 +298,618 @@ class MultiPhaseMultiComponentFluid : public MultiFluidBase
   {
     static constexpr char const * flashModelParaFileString() { return "flashModelParaFile"; }
     static constexpr char const * phasePVTParaFilesString() { return "phasePVTParaFiles"; }
-  };
-
+  } viewKeysMultiPhaseMultiComponentFluid;
 
 protected:
-  virtual void postProcessInput() override;
 
-  virtual void initializePostSubGroups() override;
+  virtual void postProcessInput() override;
 
 private:
 
   void createPVTModels();
 
-  // phase PVT parameter filenames
+  /// Names of the files defining the viscosity and density models
   path_array m_phasePVTParaFiles;
 
+  /// Name of the file defining the flash model
   Path m_flashModelParaFile;
 
-  // number of entries corrosponds to number of phases
-  std::vector< std::shared_ptr< PVTProps::PVTFunction const > > m_phaseDensityFuns;
-  std::vector< std::shared_ptr< PVTProps::PVTFunction const > > m_phaseViscosityFuns;
+  /// Index of the liquid phase
+  localIndex m_p1Index;
 
-  std::shared_ptr< PVTProps::FlashModel const > m_flashModel;
+  /// Index of the gas phase
+  localIndex m_p2Index;
+
+  /// Pointer to the brine density model
+  std::unique_ptr< P1DENS > m_p1Density;
+
+  /// Pointer to the brine viscosity model
+  std::unique_ptr< P1VISC > m_p1Viscosity;
+
+  /// Pointer to the CO2 density model
+  std::unique_ptr< P2DENS > m_p2Density;
+
+  /// Pointer to the CO2 viscosity model
+  std::unique_ptr< P2VISC > m_p2Viscosity;
+
+  /// Pointer to the flash model
+  std::unique_ptr< FLASH > m_flash;
+
+  /// Pointer to the brine density model
+  array1d< typename P1DENS::KernelWrapper > m_p1DensityWrapper;
+
+  /// Pointer to the brine viscosity model
+  array1d< typename P1VISC::KernelWrapper > m_p1ViscosityWrapper;
+
+  /// Pointer to the CO2 density model
+  array1d< typename P2DENS::KernelWrapper > m_p2DensityWrapper;
+
+  /// Pointer to the CO2 viscosity model
+  array1d< typename P2VISC::KernelWrapper > m_p2ViscosityWrapper;
+
+  /// Pointer to the flash model
+  array1d< typename FLASH::KernelWrapper > m_flashWrapper;
 
 };
 
+// this alias will be useful in constitutive dispatch
+using CO2BrineFluid = MultiPhaseMultiComponentFluid< PVTProps::BrineCO2Density,
+                                                     PVTProps::BrineViscosity,
+                                                     PVTProps::SpanWagnerCO2Density,
+                                                     PVTProps::FenghourCO2Viscosity,
+                                                     PVTProps::CO2Solubility >;
+
+// The constructor is in the hpp for one reason: the testMultiFluid won't link if the constructor is in the cpp
+// I could not find a workaround
```

Comment:
I wonder if this could be because the explicit instantiation is at the top of the file, before the compiler can see the function definitions? Try moving the instantiation to the bottom of the file (just before the catalog registration macro). If it doesn't help, let's just leave the constructor here and I'll try to figure it out later.

Reply:
Yes great! In the code you looked at, I had applied my usual recipe when things are working on Quartz but not linking on Lassen: moving the implementation to the hpp file... But now with what you suggested, I compiled fine on Lassen with the constructor implementation in the cpp. Thanks